### PR TITLE
add unsigned validtor to pallet-nft

### PR DESCRIPTION
**Pull request by ruibinx**
{Summary}

Fixes #{issue}

**Type of change**

* [ ] Bug Fix (non-breaking change which fixes an issue)
* [ ] New Feature (non-breaking change which adds functionality)
* [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update
* [ ] This change requires a storage migration
* [ ] This change requires a benchmark

**Checklist**

* [ ] My code follows the style guidelines of this project
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [ ] My changes generate no new warnings
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] New and existing benchmarks pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules
* [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by coderfengyun** [comment]
ADD: more tests for this ocw relative feature

File: pallets/nft/src/ocw/mod.rs [hunk 1]
```diff
@@ -120,28 +114,43 @@ impl<T: Config + SendTransactionTypes<Call<T>>> Pallet<T> {
         );
         let body = body
             .replace("<data>", &hex::encode(&encoded))
-            .replace("<contract>", &hex::encode(namespace));
+            .replace("<contract>", &hex::encode(&namespace));
+        return body;
+    }
 
+    pub(super) fn ocw_validate_etherum_token_owner(
```
Comment (coderfengyun) [inline pallets/nft/src/ocw/mod.rs:121]: 这个方法可以将发请求处理响应，及验证是否在links中。

分成这两个部分。

File: pallets/nft/src/tests.rs [hunk 2]
```diff
@@ -290,3 +317,50 @@ fn should_claim() {
...
+            "GoogleBot (compatible; ParamiWorker/1.0; +http://parami.io/worker/)".into(),
+        )],
+        body: body.into(),
+        response: Some(res.into()),
+        ..Default::default()
+    }
+}
+
+#[test]
+fn test_ocw_validate_etherum() {
```
Comment (coderfengyun) [inline pallets/nft/src/tests.rs:337]: should_success_when_validate_ethereum

File: pallets/nft/src/tests.rs [hunk 3]
```diff
@@ -73,6 +82,24 @@ fn should_fail_when_importing() {
     });
 }
 
+#[test]
+fn should_fail_when_did_not_linked_network() {
```
Comment (coderfengyun) [inline pallets/nft/src/tests.rs:86]: 应该加一个should_fail_when_did_linked_another_ethereum_address

Comment (coderfengyun) [reply pallets/nft/src/tests.rs:86]: 这个方法叫should_fail_when_did_linked_to_non_eth_addr

File: pallets/nft/src/tests.rs (hunk 2 above)
Comment (ruibinx) [reply pallets/nft/src/tests.rs]: done

File: pallets/nft/src/ocw/mod.rs (hunk 1 above)
Comment (ruibinx) [reply pallets/nft/src/ocw/mod.rs:123]: done

File: pallets/nft/src/ocw/mod.rs [hunk 4]
```diff
@@ -120,28 +114,43 @@ impl<T: Config + SendTransactionTypes<Call<T>>> Pallet<T> {
...
-        let res = eth_abi::decode(&[abi::ParamType::Address], &res);
-        let res = res.first();
-        match res {
-            Some(abi::Token::Address(addr)) if links.contains(&addr.as_bytes().to_vec()) => Ok(()),
-            _ => Err(Error::<T>::InvalidExternalToken)?,
+        let links: Vec<U256> = links.into_iter().map(|l| U256::from(l as &[u8])).collect();
+        if links.contains(&res) {
+            Ok(())
+        } else {
+            Err(Error::<T>::InvalidExternalToken)?
```
[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/nft/src/ocw/mod.rs]: done

File: pallets/nft/src/tests.rs (hunk 3 above)
Comment (ruibinx) [reply pallets/nft/src/tests.rs:86]: 这个地方是在测试extrinsic的port接口，验证的是当前提交的网络，用户是否已经bind，名字应该是准确的。

具体提交的token，是否属于用户，只能在ocw里验证

Comment (coderfengyun) [reply pallets/nft/src/tests.rs:86]: should_fail_when_did_linked_another_ethereum_address这个case目前有吗？

**Review by coderfengyun — APPROVED** [review]

**Review by coderfengyun — APPROVED** [review]

File: pallets/nft/src/tests.rs [hunk 5]
```diff
@@ -290,3 +319,212 @@ fn should_claim() {
...
+                b"contractaddress",
+                &token.to_be_bytes(),
+            );
+
+            assert_noop!(result, Error::<Test>::NotTokenOwner);
+        },
+    );
+}
+
+// TODO: we should test with response with status code 400, however, substrate doesn't support mocking status code for now.
```
[the remark this answers is not in the document]
Comment (ruibinx) [reply pallets/nft/src/tests.rs:415]: offchain的httpclient是substrate官方提供的，官方也提供了对应的mock，比较完善了，支持expect特定request，不过现在只支持response body和header的mock，status code是写死200的，见：https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/primitives/core/src/offchain/testing.rs#L307

加上status code的mock看起来不是很复杂，找时间给substrate提一个pr

[the remark this answers is not in the document]
Comment (coderfengyun) [reply pallets/nft/src/tests.rs:415]: 赞啊

**Pull request merged**